src/browser-extension/popup/BrowserPopup.tsx

```diff
@@ -1,41 +1,180 @@
+}
+
+export default class BrowserPopup extends React.Component<{}, Partial<IBrowserPopupState>> {
   constructor(props: {}) {
     super(props);
+
+    this.state = {
+      isLoading: true,
+      currentTabUrl: null,
+      annotationModePages: [],
+      isExtensionDisabled: false,
+      disabledPages: [],
+    };
+  }
+
+  loadStateFromStorage() {
+    /* Use data from chrome storage to update popup state
+     * It is assumed that there are no other sources of data;
+     * In the future:
+     *   for more sources of data, resort to a local popup Redux store or a global background page Redux store
+     */
+    if (this.state.isLoading) {
+      chromeStorage.get([
+        chromeKeys.ANNOTATION_MODE_PAGES,
+        chromeKeys.DISABLED_EXTENSION,
+        chromeKeys.DISABLED_PAGES,
+      ], (result) => {
+        this.setState({
+          isLoading: false,
+          annotationModePages: result[chromeKeys.ANNOTATION_MODE_PAGES] || [],
+          isExtensionDisabled: result[chromeKeys.DISABLED_EXTENSION] || false,
+          disabledPages: result[chromeKeys.DISABLED_PAGES] || [],
+        });
+      });
+    }
+
+  }
+
+  componentDidMount() {
+    chrome.tabs.query({ active: true, currentWindow: true }, (tabs) => {
+      const tab = tabs[0];
+      this.setState({ currentTabUrl: standardizeUrlForPageSettings(tab.url) });
+    });
+
+    this.loadStateFromStorage();
+  }
+
+  componentDidUpdate() {
+    this.loadStateFromStorage();
+  }
+
+  isExtensionDisabledForCurrentTab() {
+    return (this.state.disabledPages || []).indexOf(this.state.currentTabUrl) !== -1;
+  }
+
+  isAnnotationModeForCurrentTab() {
+    return (this.state.annotationModePages || []).indexOf(this.state.currentTabUrl) !== -1;
+  }
+
+  /*
+   * ON POPUP STATE PERSISTENCE
+   * We need to both update the storage and the local state
+   * - storage, because we need to communicate changes to other parts of the application
+   * - state, because there is no other clean method to ensure state continuity, so we can have interface transitions
+   * This method of communication with content scripts in fact proves cumbersome;
+   * we could consider moving to react-redux-chrome instead of refining it
+   */
+  handleAnnotationModeClick = (e) => {
+    const {
+      annotationModePages,
+      currentTabUrl,
+    } = this.state;
+
+    const isAnnotationMode = this.isAnnotationModeForCurrentTab();
+    if (!isAnnotationMode) {
+      let newAnnotationModePages;
+      newAnnotationModePages = [...annotationModePages, currentTabUrl];
+      this.setState({ annotationModePages: newAnnotationModePages });
+      chromeStorage.set({ [chromeKeys.ANNOTATION_MODE_PAGES]: newAnnotationModePages });
+      window.close();
+    }
+  }
+
+  handleDisabledExtensionChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const newValue = e.target.checked;
+    this.setState({ isExtensionDisabled: newValue });
+    chromeStorage.set({ [chromeKeys.DISABLED_EXTENSION]: newValue });
```

Comment:
as the `setState` and `chromeStorage.set` pair is repeated here and there, maybe it'd be good to wrap them in a function (e.g. `setPersistedState`)?